File: src/main/java/chapter4/conditional/AllFunctionExample.java
package chapter4.conditional;

import common.Log;
import common.Shape;
import io.reactivex.Observable;
import io.reactivex.Single;


public class AllFunctionExample {

    public void marbleDiagram() {
        String[] data = {"1", "2", "3", "4"};

        Single<Boolean> source = Observable.fromArray(data)
                .map(Shape::getShape)
                .all(Shape.BALL::equals);
        source.subscribe(Log::it);
    }

    public static void main(String[] args) {
        AllFunctionExample demo = new AllFunctionExample();
        demo.marbleDiagram();
    }

}

File: src/main/kotlin/chapter4/conditional/AllFunctionExample_.kt
package chapter4.conditional

import common.Log_
import common.Shape_
import io.reactivex.Observable


class AllFunctionExample_ {

    fun marbleDiagram() {
        val data: Array<String> = arrayOf("1", "2", "3", "4")

        val source = Observable.fromArray(*data)
                .map { Shape_.getShape(it) }
                .all { it == Shape_.BALL }
        source.subscribe { it -> Log_.it(it) }
    }

}


fun main(args: Array<String>) {
    val demo = AllFunctionExample_()
    demo.marbleDiagram()
}
